### ISSUE TITLE
RFC: Dealing with pulls that are not ready for QA

### ISSUE THREAD

**Issue by evannoronha** [iFixit/pulldasher#87]
We're running into circumstances where a pull ends up in the QA column even when it's not ready for QA. This happens in a couple different situations:
- Pull needs changes merged in
- Pull needs approval from ops
- QA needs to be done by a specific person outside QA/coders (i.e. Jeff Stephens QAing Admin pages)
- QA instructions are missing or don't fully test the new code path

This is problematic for QA because it keeps pulls in the QA column even though they fall into one of the above categories and cannot be QA'd. As we add more testers, it becomes more inefficient to have every tester look over a pull that they don't have the tools to sign off on, and it would be nice to draw attention to the pulls that need work or comments.

Proposed solution: Add a qa_block flag to Pulldasher. This flag would be added when a pull shouldn't be in the pool of pulls for testers to pick from. The pull would stay in the QA column on Pulldasher, but be moved to the bottom of the queue with a visual reminder (box shading or an icon) for testers not to open and for authors to investigate further.

Thoughts?

@iFixit/coders

**Comment by sterlinghirsh** [comment]
Why not just dev_block it?

**Comment by BaseInfinity** [comment]
Hmm all good points, I think we might be able to get away with using the `dev_block` label for things that are blocked by the developer

**Comment by lexahall** [comment]
This is something that I've wanted since I started working here, but didn't think was completely necessary when we only had 1-2 active testers. Now that we've expanded to 5 testers, I think adding a qa_block would help to streamline the QA process and could significantly increase our efficiency. I spend a lot of time looking at same pulls over and over again, only to see that they still can't be QA'd yet. Additionally, I think it would be nice for pull owners to have a visual representation of which of their pulls are just waiting on QA and which are still in need of their attention before QA can begin. 

I'm open to different ideas on how to actually implement this, but I think @evannoronha suggestion is good.

**Comment by djmetzle** [comment]
+1 for a `qa_block` label.

**Comment by lexahall** [comment]
> Why not just dev_block it?
> Hmm all good points, I think we might be able to get away with using the dev_block label for things that are blocked by the developer

I don't think that it makes sense to dev_block in the cases that Evan and I are talking about. 
For one, dev_blocking a pull removes it from the CR column, which really sucks if it just needs more instructions added for QA. I don't want to hold a pull up and prevent it from getting CR just because it isn't quite ready for QA. 
Also, qa_blocking wouldn't necessarily imply that more work is required from the dev. In the case of a lack of instructions / missing knowledge on how to actually adequately test the pull, sure - the dev needs to do a bit more work. But often when I want to qa_block something it's just because someone downstairs needs to QA it first, or something similar. In such cases, dev_blocking really wouldn't make sense and would probably just slow down the whole process and frustrate everyone. 

More than anything, it would just be really helpful for us QA's to be able to distinguish between the pulls that are ready for our attention and the ones that are not.

**Comment by addison-grant** [comment]
> Pull needs changes merged in

If you just mean merging in master, I think it's usually ok for QA to do this and then continue testing

> Pull needs approval from ops

This makes sense, sort of. I think QA can be partially responsible for setting up the testing environment for Ops, but I also think we should try to do some testing as well. Maybe we could have a `op_qa` label

> QA needs to be done by a specific person outside QA/coders (i.e. Jeff Stephens QAing Admin pages)

This falls under the Ops QA domain, but also includes TW QA, I guess.

> QA instructions are missing or don't fully test the new code path

I am sort of on board with this use case. I think it's better to meet with other QAs or devs to figure out how to test this, but that doesn't always work, so when it doesn't work, I guess `qa_block` would be appropriate.

**Comment by BaseInfinity** [comment]
When I think dev block, i think that the pull is blocked on an action required by the developer. Maybe we are missing functionality for the times a specific person is required to CR or QA. If we introduce a qa_block, do we need to also introduce a cr_block or would it be implied the same way dev_block is? :
> Pull needs changes merged in
This seems like a dev_block (depending on bad the merge conflicts are)
> Pull needs approval from ops
I could see qa_block working here
> QA needs to be done by a specific person outside QA/coders (i.e. Jeff Stephens QAing Admin pages)
could see qa_block working here
> QA instructions are missing or don't fully test the new code path
Seems like a dev_block would work here

I think we should try and get better at training our current QA to learn how to QA downstairs and dev ops stuff but that's a diff conversation.

Adding a label and filtering is easy enough though and you guys are the ones in the thick of QA so I'm totally down for getting this implemented since the risk/reward is almost non existent (adding labels and filters to pulldasher is super easy). This could be a great learning project too for adding support for qa_labels to pulldasher.

**Comment by danielj41** [comment]
A `qa_block` label seems reasonable to me.

Sort of related: I'd like the ability to "hide" a pull request so that I no longer see it. I think that's a more versatile solution to this category of problem.
  * These could still be visible under a "hidden pulls" dropdown so that no one entirely loses track of them.

**Comment by BaseInfinity** [comment]
One thing to mention if we want feature parity is would we need a QA Blocked column as well?

**Comment by lexahall** [comment]
@agrant:

> I am sort of on board with this use case. I think it's better to meet with other QAs or devs to figure out how to test this, but that doesn't always work, so when it doesn't work, I guess qa_block would be appropriate.

I often see QA instructions along the lines of "I'm not sure what to do here yet, I'll look into it". And sure, I agree it's best to meet with people and get it worked out. But that often doesn't happen when we are in rush, or if the pull owner is unavailable. I think it would be nice to be able to qa_block just to keep things more organized and also to remind us that we still need to sit down and have that discussion with other QAs/devs to figure out what to do. 

@BaseInfinity:
> If we introduce a qa_block, do we need to also introduce a cr_block or would it be implied the same way dev_block is (if something is qa_blocked, don't QA. If something is dev_blcoked, don't CR)? :

Do you think that a cr_block would be helpful? I think that dev_block pretty much covers CR issues. I think of any situations where you couldn't CR a pull that shouldn't be dev_blocked. I think QA and CR are different, and QA sometimes requires a secondary level of preparation so qa_block makes sense, while cr_block desn't really. 

> One thing to mention if we want feature parity is would we need a QA Blocked column as well?

I'm not sure a QA blocked column would be necessary. Evan was just suggesting we move the pulls to the bottom of the QA column and indicate somehow that they are blocked. If you're worried about feature parity, maybe we can name it something other than 'block'. 
I'm also okay with another column though.

**Comment by danielbeardsley** [comment]
> When I think dev block, i think that the pull is blocked on an action required by the developer. 

@BaseInfinity -- you hit it on the head. A qa_block that means "remove this from the QA column" is dangerous cause it doesn't bring it to anyone's attention, it merely hides it. dev block means action is needed by the developer; I feel that the reasons mentioned above would be better taken care of with more narrow focus.

QA instructions of "I'm not sure what to do here yet, I'll look into it" == dev block 
Merging needed == dev block
Pull needs approval == (deploy block, or dev block, depending on if we need ops for QA or spec clarification)
Outside QA needed == (same as above, though maybe QAing label is useful here)
Missing QA instructions == dev block

Dev block says "hey, dev, look at this" while a QA block would just hide it. I'm not totally opposed, but I worry about a feature who's goal is to allow us to hide / ignore more stuff.

**Comment by danielbeardsley** [comment]
> I'd like the ability to "hide" a pull request so that I no longer see it.

@danielj41 In what situations would hiding a pull be helpful? Maybe we can come up some automatic way to hide pulls that no longer require your attention? The most common reason I want to hide things is when I've done review on a pull and am expecting a response... in that case, maybe dev block is warranted?

**Comment by andyg0808** [comment]
@danielbeardsley The biggest use-case I can see for hiding a pull is when I have a pull that's based on another pull, and so I don't want people working on it until the basis pull goes out. In that case, though, maybe we want to support automatically hiding such pulls in Pulldasher until their parent pull goes out?

**Comment by danielj41** [comment]
> @danielj41 In what situations would hiding a pull be helpful? Maybe we can come up some automatic way to hide pulls that no longer require your attention?

It would just help me visually parse the CR column. For example, if there are 10 pulls in the CR column, I'd hide 5 of them. Then after I finish the first 5, I'd unhide them. Right now, I have to visually parse the whole column every time I look at it.

**Comment by lexahall** [comment]
> Dev block says "hey, dev, look at this" while a QA block would just hide it. I'm not totally opposed, but I worry about a feature who's goal is to allow us to hide / ignore more stuff.

I don't think that the intention of a QA block would be to hide it - at least that's not what I'm suggesting. I think that it would function similarly to a dev_block, and would still mean "hey, look at this" but for different reasons than a dev_block. Traditionally we have used dev_block to mean that code changes are necessary. In these cases, it makes sense to move the pull out of both the QA and CR column. 
However, if a pull is just not ready for QA, that doesn't necessarily mean it isn't ready for CR. I really don't like the idea of dev_blocking a pull and stopping it from getting CR, just because we don't know what to do for QA. Especially because, when we don't have a clear testing strategy, the more eyes we can get on the pull the better. It often happens that someone CRing the pull will offer up some suggestions for QA. 

Deploy block also doesn't really help in this situation because it doesn't move the pull out of the QA column, and it doesn't actually make sense. The pull can't be deployed until it's got it's QA anyways, so adding a deploy block doesn't mean anything. 

My thought was that QA block would not hide a pull. Instead, it would make it easier to identify which pulls are ready for QA and which still need work first. Rather than encouraging people to ignore these pulls, I was hoping it would serve as a reminder that their pull is not yet ready. 
At the moment, we just leave these pulls sitting in the QA column, where I think it is easy for the pull author to forget about them.

**Comment by kimarre** [comment]
I'm not a huge fan of the qa_block. I think it adds more clutter to Pulldasher when we could instead alter the way we use existing tools to achieve similar results. 

I think that most of the problems mentioned here should be solved by either dev_blocking ("needs QA instructions," needs merge) or in the case of QA from specific people/operations the dev should just add the `QAing` label themselves until it's ready. That way there's visual indication to testers to not bother opening that pull. 

I think it just comes down to the devs being more conscious of keeping things out of the columns they shouldn't be in. We should work on how to better we're using the tools we already have.

Alternatively, maybe instead of a special qa_block flag/column, we could just make it a label that works similarly to the `QAing` but instead replaces the eye icon with an X or something.

**Comment by lexahall** [comment]
@kim yeah, I was actually thinking about using something in place of the eye icon. Maybe just a red X would work.

**Comment by copperwall** [comment]
Sorting is mostly just giving a pull a number based on the pull's criteria, so if the return value from the pull's sort function is higher, it'll be placed higher than pull's that have a smaller value.

We could do something to penalize the score by a lot if it has a certain label. The happens in `columns.js` in the `views` directory.

**Comment by entendu** [comment]
The situations where we can't move forward seem to be when we're waiting on another department or person to sign off. What about `external_block`, with an "Externally Blocked" column?

(examples from above: Jeff Stephens, tech writing, ops...)

**Comment by lexahall** [comment]
Is anyone opposed to just adding an `external-block` or `qa-block` label that would add the QAing eye to a pull in pulldasher but styled differently (maybe just red). That way QA's can easily see which pulls are blocked, without adding any new columns or dramatically changing the way we have pulldasher set up now.

**Comment by solflare** [comment]
Should this be closed? Also did we decide on `qa-blocked` or is it still to be decided? I like `external-block` a bit more because it's clear that something is stopping progress on the issue that might not necessarily be qa related.

**Comment by lexahall** [comment]
I'm happy with external-block. I think we can close this as long as there aren't objections.

**Comment by lexahall** [comment]
Then I can open up a separate issue to implement the external-block.

**Comment by andyg0808** [comment]
No objections to `external_block` from me (but let's make sure it's an underscore for consistency).

**Comment by entendu** [comment]
@lexahall do it up!

**Comment by lexahall** [comment]
Issue opened here: https://github.com/iFixit/pulldasher/issues/99
I'm going to close this. 
I'm definitely open to suggestion though if you guys aren't happy with the icon. Just comment on #99
